### PR TITLE
Refactor IELE semantics to separate out logic for initializing accounts/storage/code/blockhashes

### PR DESCRIPTION
**Pull request by dwightguth**
In order to prepare the way for the IELE semantics to be used as the VM for a Cardano testnet, we refactor the semantics so that we no longer access account state or blockhashes except via specific klabels which load the relevant account into the accounts cell. We also configure the semantics so that some rules can be different depending on whether we are executing the semantics as a standalone or in Mantis.

The following changes are included in this PR:
* add standalone and node tags to pandoc so that we can have different rules depending on whether we are running a Mantis VM server.
* refactor the activeAccounts map into a set and use strictness so that we don't need to separately maintain a boolean for whether an account is empty or not.
* refactor `<storage>` cell so that we use `#lookupStorage` to compute the value of a storage key not in the storage map.
* refactor `<code>` cell so that code for an account can be missing and be computed by `#lookupCode`.
* refactor `iele.balance`, and `iele.extcodesize` so that an account can be missing and computed by `#newAccount`.
* fix bug in gas model of `sstore` where refunds were awarded for growing storage instead of for shrinking it.
* make process .build/vm/iele-vm which executes JSON tests using ProtoBuf API
* update JSON tests to more thoroughly test outputs

### CODE REVIEW

File: iele-node.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+           ...
+         </accounts>
+      requires notBool ACCT in ACCTS
+
+    syntax Int ::= #getStorageData(Int, Int) [function, hook(MANTIS.getStorageData)]
+ // --------------------------------------------------------------------------------
+    rule <k> #lookupStorage(ACCT, INDEX) => . ... </k>
+         <account>
+           <acctID>  ACCT                                                         </acctID>
+           <storage> STORAGE => STORAGE [ INDEX <- #getStorageData(ACCT, INDEX) ] </storage>
```
Comment (yzhang90) [inline iele-node.md]: why updating the storage here? What is "#getStorageData" supposed to do?

**Comment by dwightguth** [comment]
I have tested this by running make test and make iele-test-node, so the tests are passing both in standalone and in node format. This still does not include networking, but it is now ready for review.

@bmmoore @theo25 @yzhang90 if you could take a look that would be good.

I am going to get started tomorrow on integrating the networking code.

File: iele.md [hunk 2]
```diff
@@ -406,6 +406,129 @@ Some instructions require an argument to be interpreted as an address (modulo 16
...
+           <storage> STORAGE => STORAGE [ INDEX <- 0 ] </storage>
+           ...
+         </account>
+      requires notBool INDEX in_keys(STORAGE)
+```
+
+-   `#lookupCode` loads the `<code>` cell with the code for a particular account.
+
+```{.k .uiuck .rvk .standalone .node}
+    syntax InternalOp ::= #lookupCode ( Int )
```
Comment (theo25) [inline iele.md]: where are the rewrite rules for this in case of standalone?

Comment (dwightguth) [reply iele.md]: They are not necessary as lookupCode only applies if the code is .Contract, which is not a well formed contract and is therefore not actually ever the code of any account; we simply reserve it to refer to an account that the node has not yet lazily computed code for.

**Comment by dwightguth** [comment]
@theo25 please re review

File: ethereum.md [hunk 3]
```diff
@@ -17,14 +23,18 @@ module ETHEREUM-SIMULATION
...
 
+```{.k .uiuck .rvk .node}
+    imports IELE-NODE
+```
+
 A IELE simulation is a list of IELE commands.
 Some IELE commands take a specification of IELE state (eg. for an account or transaction).
 
-```{.k .uiuck .rvk}
+```{.k .uiuck .rvk .standalone }
```
Comment (yzhang90) [inline ethereum.md]: There is no IELESimulation defined in .node version? but In .node version, the k cell in the configuration still accepts IELESimulation, right?

File: ethereum.md [hunk 4]
```diff
@@ -86,7 +96,7 @@ To do so, we'll extend sort `JSON` with some IELE specific syntax, and provide a
 -   `#adjustGas` fakes the gas usage of the transaction since the EVM-to-IELE conversion does not preserve gas usage.
 -   `finishTx` is a place-holder for performing necessary cleanup after a transaction.
 
-```{.k .uiuck .rvk}
+```{.k .uiuck .rvk .standalone .node}
```
Comment (yzhang90) [inline ethereum.md]: If I understand it correctly, the entry point for the .node version should be `runVM`, why include startTx, loadTx, and etc in the .node version?

File: ethereum.md [hunk 5]
```diff
@@ -182,14 +196,14 @@ To do so, we'll extend sort `JSON` with some IELE specific syntax, and provide a
          </message>
 
     rule <k> #adjustGas => . ... </k> <checkGas> true </checkGas>
+    rule STATUS:Int ~> #adjustGas => #load %0 STATUS ~> #adjustGas
 
     syntax IELECommand ::= "#finishTx"
  // ----------------------------------
-    rule <k> #exception _ ~> #finishTx => #popCallStack ~> #popWorldState ~> #popSubstate ... </k>
-    rule <k> #revert _    ~> #finishTx => #popCallStack ~> #popWorldState ~> #popSubstate ~> #refund GAVAIL ... </k> <gas> GAVAIL </gas>       
+    rule <k> #exception STATUS ~> #finishTx => #popCallStack ~> #popWorldState ~> #popSubstate ~> #load %0 STATUS ... </k>
```
Comment (yzhang90) [inline ethereum.md]: Is this change (loading STATUS to the register) not listed in the PR summary?

File: iele-semantics-plugin/MANTIS.ml [hunk 6]
```diff
@@ -0,0 +1,35 @@
+open Constants
+open Constants.K
+open World
+
+module Cache = Caching.Make(World.InMemoryWorldState)
+
+let hook_getBalance c _ _ config _ = match c with
```
Comment (yzhang90) [inline iele-semantics-plugin/MANTIS.ml]: Could you kindly remind me what are the five arguments to the function hook_getBalance?

File: ethereum.md (hunk 5 above)
Comment (dwightguth) [reply ethereum.md]: updated

File: ethereum.md (hunk 4 above)
Comment (dwightguth) [reply ethereum.md]: I have disabled more of the code in node that is not needed.

File: iele-semantics-plugin/MANTIS.ml (hunk 6 above)
Comment (dwightguth) [reply iele-semantics-plugin/MANTIS.ml]: arguments, klabel, return sort, configuration, fresh function. This is the signature of all hooks in the OCAML backend, but most of these are not used by these hooks.

File: iele.md [hunk 7]
```diff
@@ -406,6 +406,129 @@ Some instructions require an argument to be interpreted as an address (modulo 16
...
+         <activeAccounts> ACCTS </activeAccounts>
+      requires notBool ACCT in ACCTS
+```
+
+```{.k .uiuck .rvk .standalone}
+    rule <k> #loadAccount ACCT => . ... </k>
+         <activeAccounts> ACCTS (.Set => SetItem(ACCT)) </activeAccounts>
+         <accounts>
+           ( .Bag
+          => <account>
```
Comment (yzhang90) [inline iele.md]: If the ACCT need to be created, should the balance, code, nonce cell be initialized?

File: data.md [hunk 8]
```diff
@@ -5,7 +5,7 @@ IELE uses arbitrary-precision integers, and sometimes also bytes (8 bit words).
 Here we provide the arithmetic of these words, as well as some data-structures over them.
 Both are implemented using K's `Int`.
 
-```{.k .uiuck .rvk}
+```{.k .uiuck .rvk .standalone .node}
```
[the remark this answers is not in the document]
Comment (dwightguth) [reply data.md]: I don't really understand the lua code very well, if you're interested in this, can you make a patch?

**Comment by dwightguth** [comment]
I believe I have addressed all the issues with the exception of the Lua one, which Brandon says he will make an issue for. Since we want the IOHK people to be able to run this, I am going to merge this, but if you wanted anything else to be included message me and I'll make a followup PR asap.

**Pull request merged**